Name - Chetan Singh<br>
Email - chetansingh156578@gmail.com<br>
GitHub - https://github.com/chetansingh156578

In [1]:
# importing pandas library for data handling and manipulation

import pandas as pd

In [2]:
# Reading the json file and converting it to a pandas dataframe for easy data manipulation

df = pd.read_json("C:\\Users\\CHETAN SINGH\\Downloads\\domainTags2.json")

In [3]:
# Checking the head of our dataframe

df.head(5)

_id  \
0  {'$oid': '5d09dfd68501d21d7d729bb1'}   
1  {'$oid': '5d09dfd78501d21d7d729bbb'}   
2  {'$oid': '5d09dfd78501d21d7d729bbc'}   
3  {'$oid': '5d09dfd78501d21d7d729bbe'}   
4  {'$oid': '5d09dfd78501d21d7d729bbf'}   

                                             tags               companyName  \
0                            [finance, insurance]  [21st century insurance]   
1                                 [3d technology]             [3d networks]   
2  [3d technology, augmented reality, e-commerce]                   [3dphy]   
3          [android, enterprise software, mobile]                     [3lm]   
4  [automotive, electronics, enterprise software]                      [3m]   

   domainName                                     crunchbase_url  \
0        21st  https://www.crunchbase.com/organization/21st-c...   
1  3dnetworks  https://www.crunchbase.com/organization/3d-net...   
2       3dphy  https://www.crunchbase.com/organization/3dphy-...   
3         3lm  https://www.crunchbase.com/organization/3lm?ut...   
4          3m  https://www.crunchbase.com/organization/3m?utm...   

            TYPE                              UpdatedOn  \
0  Product-based  {'$date': '2020-01-18T03:39:34.845Z'}   
1  Product-based  {'$date': '2020-01-20T15:10:33.852Z'}   
2  Product-based  {'$date': '2020-01-16T11:44:30.116Z'}   
3  Product-based  {'$date': '2020-01-21T12:50:14.767Z'}   
4  Product-based  {'$date': '2020-01-17T17:56:53.792Z'}   

                                         companyDesc      companyWebsite  \
0  since 1958, 21st century insurance has been de...        www.21st.com   
1                                                NaN  www.3dnetworks.com   
2  3dphy provides a range of 3d virtual and augme...       www.3dphy.com   
3  3lm has already signed deals to provide its te...         www.3lm.com   
4  3m operates as a diversified technology compan...          www.3m.com   

                employeeNumber                                   foundedDate  \
0  {'max': 10000, 'min': 5001}   {'$date': {'$numberLong': '-378658800000'}}   
1                          NaN         {'$date': '2000-01-01T09:00:00.000Z'}   
2       {'max': 50, 'min': 11}         {'$date': '2014-11-03T09:00:00.000Z'}   
3       {'max': 50, 'min': 11}         {'$date': '2010-07-01T09:00:00.000Z'}   
4     {'max': 10001, 'min': 0}  {'$date': {'$numberLong': '-2145884400000'}}   

                                       domain_groups  \
0  {'Fintech/ Payments': 0.825334256973911, 'Insu...   
1                                                NaN   
2                                                NaN   
3             {'Developer Tools': 0.617930901407716}   
4  {'Hardware': 0.6879067738004571, 'Automotive':...   

                                         fundingList  
0                                                NaN  
1                                                NaN  
2  [{'fundingDate': {'$date': '2014-12-18T09:00:0...  
3  [{'fundingDate': {'$date': '2010-07-01T09:00:0...  
4                                                NaN

In [4]:
# Getting info about the dataframe

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7462 entries, 0 to 7461
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   _id             7462 non-null   object
 1   tags            7397 non-null   object
 2   companyName     7462 non-null   object
 3   domainName      7462 non-null   object
 4   crunchbase_url  7462 non-null   object
 5   TYPE            6097 non-null   object
 6   UpdatedOn       7428 non-null   object
 7   companyDesc     6247 non-null   object
 8   companyWebsite  7336 non-null   object
 9   employeeNumber  6604 non-null   object
 10  foundedDate     6603 non-null   object
 11  domain_groups   4406 non-null   object
 12  fundingList     2140 non-null   object
dtypes: object(13)
memory usage: 758.0+ KB


So from above one observation we can make is that there seems to be some values missing in our dataset

In [5]:
df.shape

(7462, 13)

So from above we can see that our dataset contains 7462 rows and 13 columns or fields which we have already get from df.info()

So let's once look at the data dictionary of the fiels provided in the prblem statement

- tags: the domains company operates in

- companyName: synonyms of the company

- domainName: extracted from the company website’s URL.

- companyDesc: a short description of the company's operations

- fundingList: details regarding the company’s funding rounds

- crunchbase_url: URL to the company’s Crunchbase page

- employeeNumber: number of employees in the company at the time data was extracted

- foundedDate: date the company was founded on

- companyWebsite: URL to company’s website

- domain_groups: listing out possible groups the company may belong to. (this is our internal classification, may not be accurate)

- TYPE: another internal classification (for product, and service-based companies), may not be accurate

Also let's note down the given problem statements one by one

## Problem Statement 1

given a company name (similar to ones present in companyName), return the close competitors of the company, using any of the fields (or combination of fields) from the dataset. For example, competitors of Swiggy would be Zomato, FoodPanda, DoorDash, etc. Since all of them operate in similar domains and offer similar services.


Extracting company names and tags in the form of pandas.series and later on converting them to python list() for more liberation in data manipulation

In [6]:
serie1, serie2 = df['companyName'], df['tags']

In [7]:
list1 = list(serie1) #Converting serie1 from pandas.Series to list
list2 = list(serie2) # Converting serie2 from pandas.Series to list


So let's say the given company name is zomato so i will first search up 'zomato' and get's its location in the database

In [8]:
for i in range(len(list1)):
    if 'zomato' in list1[i]:
        print(i)

5512


I've got the index or location of the company zomato which is 5512, now i will get all the details of the company mainly the tags of the company

In [9]:
df.iloc[5512]

_id                            {'$oid': '5d09e0258501d21d7d72b0fe'}
tags              [food delivery, hospitality, reservations, res...
companyName                                                [zomato]
domainName                                                   zomato
crunchbase_url    https://www.crunchbase.com/organization/zomato...
TYPE                                                  Product-based
UpdatedOn                     {'$date': '2020-02-06T14:48:12.523Z'}
companyDesc       zomato offers a search app that provides detai...
companyWebsite                                      www.zomato.com/
employeeNumber                           {'max': 5000, 'min': 1001}
foundedDate                   {'$date': '2008-07-10T09:00:00.000Z'}
domain_groups                                                    {}
fundingList       [{'fundingDate': {'$date': '2019-03-01T09:00:0...
Name: 5512, dtype: object

Let's make a list of the competitors of the comapny based of it's first tag which should be the main service provided by the company and will give us the close competitors of the company in that domain

In [10]:
competitors = []
main_domain = df.iloc[5512]['tags'][0]
main_domain

'food delivery'

In [11]:
import numpy as np

In [12]:
for i in range(len(list2)):
    if list2[i]:
        if str(type(list2[i])) == "<class 'float'>":   # In order to avoid np.nan values
            pass
        else:
            if main_domain in list2[i]:
                competitors.append(i)

In [13]:
competitors

[487,
 1121,
 1481,
 1632,
 1686,
 1804,
 1951,
 2261,
 2849,
 3820,
 3826,
 4548,
 4780,
 4789,
 4919,
 5512,
 5910,
 6190,
 6472,
 6861,
 7047,
 7268,
 7426]

In [14]:
df.iloc[competitors]

_id  \
487   {'$oid': '5d09dfdc8501d21d7d729da0'}   
1121  {'$oid': '5d09dfe38501d21d7d72a027'}   
1481  {'$oid': '5d09dfea8501d21d7d72a290'}   
1632  {'$oid': '5d09dfec8501d21d7d72a317'}   
1686  {'$oid': '5d09dfeb8501d21d7d72a2f1'}   
1804  {'$oid': '5d09dfec8501d21d7d72a318'}   
1951  {'$oid': '5d09dfeb8501d21d7d72a2f0'}   
2261  {'$oid': '5d09dff48501d21d7d72a56b'}   
2849  {'$oid': '5d09dffa8501d21d7d72a700'}   
3820  {'$oid': '5d09e0048501d21d7d72a9e6'}   
3826  {'$oid': '5d09e0078501d21d7d72aa7f'}   
4548  {'$oid': '5d09e0128501d21d7d72ad05'}   
4780  {'$oid': '5d09e0148501d21d7d72ad7a'}   
4789  {'$oid': '5d09e0178501d21d7d72ae25'}   
4919  {'$oid': '5d09e01b8501d21d7d72af0b'}   
5512  {'$oid': '5d09e0258501d21d7d72b0fe'}   
5910  {'$oid': '5d09e02c8501d21d7d72b2cd'}   
6190  {'$oid': '5d09e0328501d21d7d72b40e'}   
6472  {'$oid': '5e4678ee09a067622f5970b9'}   
6861  {'$oid': '5e46792509a067622f5972bb'}   
7047  {'$oid': '5e4a72196a9636171e7a6162'}   
7268  {'$oid': '5e4a72106a9636171e7a6124'}   
7426  {'$oid': '5e4a721c6a9636171e7a617b'}   

                                                   tags  \
487   [digital marketing, e-commerce, food and bever...   
1121  [catering, food and beverage, food delivery, f...   
1481       [e-commerce, food delivery, food processing]   
1632  [delivery, food and beverage, food delivery, f...   
1686              [delivery, e-commerce, food delivery]   
1804  [e-commerce, e-commerce platforms, food and be...   
1951  [delivery, e-commerce platforms, food delivery...   
2261  [cooking, e-commerce, food and beverage, food ...   
2849  [food delivery, food processing, information t...   
3820                 [food and beverage, food delivery]   
3826  [delivery, food delivery, logistics, same day ...   
4548  [e-commerce, food and beverage, food delivery,...   
4780  [food delivery, food processing, internet, sna...   
4789                  [food delivery, retail, shopping]   
4919                 [apps, food delivery, mobile apps]   
5512  [food delivery, hospitality, reservations, res...   
5910  [e-commerce, financial services, fintech, food...   
6190       [consumer, food and beverage, food delivery]   
6472  [e-commerce, food delivery, health care, organ...   
6861                 [food and beverage, food delivery]   
7047              [catering, e-commerce, food delivery]   
7268  [food and beverage, food delivery, food proces...   
7426            [b2b, food and beverage, food delivery]   

                                            companyName        domainName  \
487                              [avanti schools trust]            avanti   
1121                                          [cookifi]           cookifi   
1481                 [faasos, faasos's food services .]            faasos   
1632                                        [freshmenu]         freshmenu   
1686                           [foodpandain, foodpanda]         foodpanda   
1804                [freshtohomecom, wwwfreshtohomecom]       freshtohome   
1951                                          [foodora]           foodora   
2261                                        [innerchef]         innerchef   
2849                                [lineage logistics]  lineagelogistics   
3820                                         [ox verte]           oxverte   
3826                                        [postmates]         postmates   
4548                                        [spoonshot]         spoonshot   
4780                                           [swiggy]            swiggy   
4789  [tesco, tesco hindustan service centre, tesco ...             tesco   
4919                                         [ubereats]          ubereats   
5512                                           [zomato]            zomato   
5910                   [gojek tech, gojek, gojek india]             gojek   
6190                                         [zume inc]              zume   
6472                                             [frsh

So here is a list of competitors of zomato. Let's get the company names separately

In [15]:
df.iloc[competitors]['companyName']

487                                [avanti schools trust]
1121                                            [cookifi]
1481                   [faasos, faasos's food services .]
1632                                          [freshmenu]
1686                             [foodpandain, foodpanda]
1804                  [freshtohomecom, wwwfreshtohomecom]
1951                                            [foodora]
2261                                          [innerchef]
2849                                  [lineage logistics]
3820                                           [ox verte]
3826                                          [postmates]
4548                                          [spoonshot]
4780                                             [swiggy]
4789    [tesco, tesco hindustan service centre, tesco ...
4919                                           [ubereats]
5512                                             [zomato]
5910                     [gojek tech, gojek, gojek india]
6190          

<font color='blue'>So our solution to the first problem is done here</font>

## Problem Statement 2

Given a domain (any one of the domains in ‘tags’ field), present the companies with the fastest, and the slowest growth (hint: it may be based on the fundingList, foundedDate, etc)

Let's assume the given tag, i'll go with 'pharmaceutical'.
So let's understand what do we infer from growth of a company. First let's print a few tuples of the dataset.

In [21]:
df.head(5)

_id  \
0  {'$oid': '5d09dfd68501d21d7d729bb1'}   
1  {'$oid': '5d09dfd78501d21d7d729bbb'}   
2  {'$oid': '5d09dfd78501d21d7d729bbc'}   
3  {'$oid': '5d09dfd78501d21d7d729bbe'}   
4  {'$oid': '5d09dfd78501d21d7d729bbf'}   

                                             tags               companyName  \
0                            [finance, insurance]  [21st century insurance]   
1                                 [3d technology]             [3d networks]   
2  [3d technology, augmented reality, e-commerce]                   [3dphy]   
3          [android, enterprise software, mobile]                     [3lm]   
4  [automotive, electronics, enterprise software]                      [3m]   

   domainName                                     crunchbase_url  \
0        21st  https://www.crunchbase.com/organization/21st-c...   
1  3dnetworks  https://www.crunchbase.com/organization/3d-net...   
2       3dphy  https://www.crunchbase.com/organization/3dphy-...   
3         3lm  https://www.crunchbase.com/organization/3lm?ut...   
4          3m  https://www.crunchbase.com/organization/3m?utm...   

            TYPE                              UpdatedOn  \
0  Product-based  {'$date': '2020-01-18T03:39:34.845Z'}   
1  Product-based  {'$date': '2020-01-20T15:10:33.852Z'}   
2  Product-based  {'$date': '2020-01-16T11:44:30.116Z'}   
3  Product-based  {'$date': '2020-01-21T12:50:14.767Z'}   
4  Product-based  {'$date': '2020-01-17T17:56:53.792Z'}   

                                         companyDesc      companyWebsite  \
0  since 1958, 21st century insurance has been de...        www.21st.com   
1                                                NaN  www.3dnetworks.com   
2  3dphy provides a range of 3d virtual and augme...       www.3dphy.com   
3  3lm has already signed deals to provide its te...         www.3lm.com   
4  3m operates as a diversified technology compan...          www.3m.com   

                employeeNumber                                   foundedDate  \
0  {'max': 10000, 'min': 5001}   {'$date': {'$numberLong': '-378658800000'}}   
1                          NaN         {'$date': '2000-01-01T09:00:00.000Z'}   
2       {'max': 50, 'min': 11}         {'$date': '2014-11-03T09:00:00.000Z'}   
3       {'max': 50, 'min': 11}         {'$date': '2010-07-01T09:00:00.000Z'}   
4     {'max': 10001, 'min': 0}  {'$date': {'$numberLong': '-2145884400000'}}   

                                       domain_groups  \
0  {'Fintech/ Payments': 0.825334256973911, 'Insu...   
1                                                NaN   
2                                                NaN   
3             {'Developer Tools': 0.617930901407716}   
4  {'Hardware': 0.6879067738004571, 'Automotive':...   

                                         fundingList  
0                                                NaN  
1                                                NaN  
2  [{'fundingDate': {'$date': '2014-12-18T09:00:0...  
3  [{'fundingDate': {'$date': '2010-07-01T09:00:0...  
4                                                NaN

Here three coulumns are of my are of interest
- employeeNumber (Gives us an estimation of the size of a company in terms of manpower)
- foundedDate (type = dictionary)
- fundingList (type = list(dictionary))

I will here again get the info about the dataset

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7462 entries, 0 to 7461
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   _id             7462 non-null   object
 1   tags            7397 non-null   object
 2   companyName     7462 non-null   object
 3   domainName      7462 non-null   object
 4   crunchbase_url  7462 non-null   object
 5   TYPE            6097 non-null   object
 6   UpdatedOn       7428 non-null   object
 7   companyDesc     6247 non-null   object
 8   companyWebsite  7336 non-null   object
 9   employeeNumber  6604 non-null   object
 10  foundedDate     6603 non-null   object
 11  domain_groups   4406 non-null   object
 12  fundingList     2140 non-null   object
dtypes: object(13)
memory usage: 758.0+ KB


Here we can see that fundingList has a lot of null values present so it won't serve as a good metric to evaluate a company's growth so using the fundingList column is out of the picture.
However employee number and foundedDate provide a promising non-null value number so i will go ahead with using these two colums

I will here cleverely invent a company growth metric which will be defined as follows:-

**metric = (employeeNumber['max']-employeeNumber['min'])*(foundedDate(Year)+foundedDate(Month)+foundedYear(Date))/100000**

So what basically this metric will do is that it will us an estimate how much a company has expanded in size ever since it was founded. The **metric** variable will be of float type. Greater the value of metric greater the growth of the company and the opposite also holds true. 

So let's first get the location of the companies that have tag 'pharmaceutical'

In [23]:
pharmaceutical = []
for i in range(len(list2)):
    if list2[i]:
        if str(type(list2[i])) == "<class 'float'>":
            pass
        else:
            if 'pharmaceutical' in list2[i]:
                pharmaceutical.append(i)
    

In [25]:
pharmaceutical[:10]

[18, 46, 75, 97, 115, 142, 184, 218, 339, 404]

So here we've got the companies with tags as pharmaceutical.
Let's also fetch the **employeeNumber** and **foundedDate** in the form of lists for better manipulation

In [26]:
serie3, serie4 = df['employeeNumber'].iloc[pharmaceutical], df['foundedDate'].iloc[pharmaceutical]

In [27]:
list3, list4 = list(serie3), list(serie4)

Now let's first get the information of our use from the list3 i.e employeeNumber(max)-employeeNumber(min)

In [28]:
for i in range(len(list3)):
    if list3[i]:
        if str(type(list3[i])) == "<class 'float'>":
            pass
        else:
            numerator1 = list3[i]['max'] - list3[i]['min']
            list3[i] = numerator1

In [29]:
list3[:10]

[4999, 10001, 10001, 39, 10001, nan, 3999, 249, 499, 499]

So here we have successfully got the first part of numerator values for our metrics

Now let's get the second part of the numerator values of our metrics

In [30]:
for i in range(len(list4)):
    if list4[i]:
        if str(type(list4[i])) == "<class 'float'>":
            pass
        else:
            if str(type(list4[i]['$date'])) == "<class 'dict'>":  # in oder to handle numberLong dates
                pass
            else:
                temp = 0
                date = list4[i]['$date'][:10]
                date = date.split('-')
                temp += int(date[0])
                temp += int(date[1])
                temp += int(date[2])
                list4[i] = temp
            
        

In [31]:
list4[:15]

[{'$date': {'$numberLong': '-1988118000000'}},
 1975,
 2015,
 2012,
 {'$date': {'$numberLong': '-694278000000'}},
 1982,
 1989,
 2007,
 2006,
 1997,
 {'$date': {'$numberLong': '-2682255600000'}},
 1988,
 1984,
 2009,
 {'$date': {'$numberLong': '-1104505200000'}}]

So here we have obtained our numerator2 also, so let's calculate the metrics

In [32]:
metrics = []

for i in range(len(list3)):
    if str(type(list3[i])) == "'<class 'float'>'" or str(type(list4[i])) == "'<class 'float'>'" or str(type(list4[i])) == "<class 'dict'>":
        metrics.append(-1)
    else:
        metrics.append((list3[i]*list4[i])/100000)

In [34]:
metrics[:20]

[-1,
 197.51975,
 201.52015,
 0.78468,
 -1,
 nan,
 79.54011,
 4.99743,
 10.00994,
 9.96503,
 -1,
 198.81988,
 99.18016,
 200.92009,
 -1,
 nan,
 99.18016,
 0.98,
 0.78,
 199.31993]

So now we've got our metrics, now we can finally choose our top and least growing companies
So let's get our top ten pharmaceutical companies in terms of growth

In [35]:
import math

In [37]:
# Removing -1 and nan values from my dataset
temp = []
for i in metrics:
    if math.isnan(i) or i==-1:
        pass
    else:
        temp.append(i)
temp[:15]

[197.51975,
 201.52015,
 0.78468,
 79.54011,
 4.99743,
 10.00994,
 9.96503,
 198.81988,
 99.18016,
 200.92009,
 99.18016,
 0.98,
 0.78,
 199.31993,
 5.0049]

In [38]:
temp.sort(reverse=True)

In [39]:
temp[:11]

[201.52015,
 200.92009,
 200.82008,
 200.62006,
 199.91999,
 199.81998,
 199.81998,
 199.61996,
 199.31993,
 199.21992,
 198.91989]

In [40]:
top10pharmaceutical = temp[:11]

In [41]:
indexoftop10pharmaceutical = []

for i in range(len(metrics)):
    if metrics[i] == -1 or math.isnan(metrics[i]):
        pass
    elif metrics[i] in top10pharmaceutical:
        indexoftop10pharmaceutical.append(i)
        

indexoftop10pharmaceutical    

[2, 13, 19, 22, 25, 26, 36, 42, 47, 54, 82]

In [42]:
ans1 = []
for i in indexoftop10pharmaceutical:
    ans1.append(pharmaceutical[i])
ans1

[75, 626, 1137, 1184, 1394, 1478, 2135, 2283, 2795, 3343, 6973]

In [43]:
df.iloc[ans1]

_id  \
75    {'$oid': '5d09dfd78501d21d7d729bda'}   
626   {'$oid': '5d09dfdb8501d21d7d729d65'}   
1137  {'$oid': '5d09dfe18501d21d7d729fa2'}   
1184  {'$oid': '5d09dfe48501d21d7d72a08f'}   
1394  {'$oid': '5d09dfe38501d21d7d72a03f'}   
1478  {'$oid': '5d09dfe98501d21d7d72a25c'}   
2135  {'$oid': '5d09dff28501d21d7d72a4dd'}   
2283  {'$oid': '5d09dff18501d21d7d72a492'}   
2795  {'$oid': '5d09dff88501d21d7d72a677'}   
3343  {'$oid': '5d09e0028501d21d7d72a93f'}   
6973  {'$oid': '5e46790809a067622f5971b9'}   

                                                   tags  \
75    [biotechnology, health care, medical, pharmace...   
626   [biopharma, biotechnology, health care, medica...   
1137  [biotechnology, health diagnostics, medical, p...   
1184  [health care, medical, personal health, pharma...   
1394  [biotechnology, dietary supplements, health ca...   
1478  [biotechnology, genetics, life science, pharma...   
2135  [biopharma, biotechnology, medical, pharmaceut...   
2283          [medical, medical device, pharmaceutical]   
2795                         [pharmaceutical, software]   
3343                      [health care, pharmaceutical]   
6973            [health care, pharmaceutical, wellness]   

                                            companyName    domainName  \
75                                             [abbvie]        abbvie   
626   [astrazeneca pharmaceuticals, astrazeneca indi...   astrazeneca   
1137                              [client roche, roche]         roche   
1184                                       [cvs health]     cvshealth   
1394                                          [covance]       covance   
1478  [eurofins it solutions india subsidiary of eur...      eurofins   
2135                                         [icon plc]       iconplc   
2283                                          [hospira]       hospira   
2795      [kmi parexel, parexel international, parexel]       parexel   
3343  [novartis, novartis company, novartis consumer...      novartis   
6973                                          [medplus]  medplusindia   

                                         crunchbase_url           TYPE  \
75    https://www.crunchbase.com/organization/abbvie...  Product-based   
626   https://www.crunchbase.com/organization/astraz...  Product-based   
1137  https://www.crunchbase.com/organization/roche-...  Product-based   
1184  https://www.crunchbase.com/organization/cvs-ca...  Product-based   
1394  https://www.crunchbase.com/organization/covanc...  Service-based   
1478  https://www.crunchbase.com/organization/eurofi...  Service-based   
2135  https://www.crunchbase.com/organization/icon-p...  Product-based   
2283  https://www.crunchbase.com/organization/hospir...  Product-based   
2795  https://www.crunchbase.com/organization/liquen...  Product-based   
3343  https://www.crunchbase.com/organization/novart...  Service-based   
6973  https://www.crunchbase.com/organization/medplus-2            NaN   

                                  UpdatedOn  \
75    {'$date': '2020-01-17T00:10:10.435Z'}   
626   {'$date': '2020-01-20T10:00:33.925Z'}   
1137  {'$date': '2020-01-14T22:00:10.427Z'}   
1184  {'$date': '2020-01-16T17:11:07.874Z'}   
1394  {'$date': '2020-01-17T08:10:04.383Z'}   
1478  {'$date': '2020-01-28T09:00:08.542Z'}   
2135  {'$date': '2020-01-27T00:35:22.545Z'}   
2283  {'$date': '2020-02-05T12:06:11.758Z'}   
2795  {'$date': '2020-01-18T14:38:26.777Z'}   
3343  {'$date': '2020-01-17T14:57:55.899Z'}   
6973  {'$date': '2020-02-14T10:40:08.946Z'}   

                                            companyDesc  \
75    abbvie is a global, research-based biopharmace...   
626   astrazeneca is global innovation-driven biopha...   
1137  roche molecular systems, inc., a molecular dia...   
1184  following its acquisition of aetna in late 201...   
1394  covance inc. provides early-and late-stage ser...   
1478  eurofins scientific is an international life s...   
2135  icon is a global provider o

So here we've got our top ten pharmaceutical companies in terms of growth calculated based on my defined metric. Similarly we can get bottom ten companies in terms of growth calculated based on my defined metric.

In [44]:
bottom10pharmaceutical = temp[-10:]

In [45]:
bottom10pharmaceutical

[0.78468,
 0.78468,
 0.78312,
 0.78234,
 0.78,
 0.77532,
 0.18279,
 0.1818,
 0.18144,
 0.17748]

In [46]:
indexofbottom10pharmaceutical = []

for i in range(len(metrics)):
    if metrics[i] == -1 or math.isnan(metrics[i]):
        pass
    elif metrics[i] in bottom10pharmaceutical:
        indexofbottom10pharmaceutical.append(i)
        

indexofbottom10pharmaceutical   

[3, 18, 23, 24, 27, 35, 46, 71, 78, 80]

In [47]:
ans2 = []
for i in indexofbottom10pharmaceutical:
    ans2.append(pharmaceutical[i])
ans2

[97, 992, 1274, 1331, 1656, 2132, 2754, 5437, 6137, 6545]

In [48]:
df.iloc[ans1]

_id  \
75    {'$oid': '5d09dfd78501d21d7d729bda'}   
626   {'$oid': '5d09dfdb8501d21d7d729d65'}   
1137  {'$oid': '5d09dfe18501d21d7d729fa2'}   
1184  {'$oid': '5d09dfe48501d21d7d72a08f'}   
1394  {'$oid': '5d09dfe38501d21d7d72a03f'}   
1478  {'$oid': '5d09dfe98501d21d7d72a25c'}   
2135  {'$oid': '5d09dff28501d21d7d72a4dd'}   
2283  {'$oid': '5d09dff18501d21d7d72a492'}   
2795  {'$oid': '5d09dff88501d21d7d72a677'}   
3343  {'$oid': '5d09e0028501d21d7d72a93f'}   
6973  {'$oid': '5e46790809a067622f5971b9'}   

                                                   tags  \
75    [biotechnology, health care, medical, pharmace...   
626   [biopharma, biotechnology, health care, medica...   
1137  [biotechnology, health diagnostics, medical, p...   
1184  [health care, medical, personal health, pharma...   
1394  [biotechnology, dietary supplements, health ca...   
1478  [biotechnology, genetics, life science, pharma...   
2135  [biopharma, biotechnology, medical, pharmaceut...   
2283          [medical, medical device, pharmaceutical]   
2795                         [pharmaceutical, software]   
3343                      [health care, pharmaceutical]   
6973            [health care, pharmaceutical, wellness]   

                                            companyName    domainName  \
75                                             [abbvie]        abbvie   
626   [astrazeneca pharmaceuticals, astrazeneca indi...   astrazeneca   
1137                              [client roche, roche]         roche   
1184                                       [cvs health]     cvshealth   
1394                                          [covance]       covance   
1478  [eurofins it solutions india subsidiary of eur...      eurofins   
2135                                         [icon plc]       iconplc   
2283                                          [hospira]       hospira   
2795      [kmi parexel, parexel international, parexel]       parexel   
3343  [novartis, novartis company, novartis consumer...      novartis   
6973                                          [medplus]  medplusindia   

                                         crunchbase_url           TYPE  \
75    https://www.crunchbase.com/organization/abbvie...  Product-based   
626   https://www.crunchbase.com/organization/astraz...  Product-based   
1137  https://www.crunchbase.com/organization/roche-...  Product-based   
1184  https://www.crunchbase.com/organization/cvs-ca...  Product-based   
1394  https://www.crunchbase.com/organization/covanc...  Service-based   
1478  https://www.crunchbase.com/organization/eurofi...  Service-based   
2135  https://www.crunchbase.com/organization/icon-p...  Product-based   
2283  https://www.crunchbase.com/organization/hospir...  Product-based   
2795  https://www.crunchbase.com/organization/liquen...  Product-based   
3343  https://www.crunchbase.com/organization/novart...  Service-based   
6973  https://www.crunchbase.com/organization/medplus-2            NaN   

                                  UpdatedOn  \
75    {'$date': '2020-01-17T00:10:10.435Z'}   
626   {'$date': '2020-01-20T10:00:33.925Z'}   
1137  {'$date': '2020-01-14T22:00:10.427Z'}   
1184  {'$date': '2020-01-16T17:11:07.874Z'}   
1394  {'$date': '2020-01-17T08:10:04.383Z'}   
1478  {'$date': '2020-01-28T09:00:08.542Z'}   
2135  {'$date': '2020-01-27T00:35:22.545Z'}   
2283  {'$date': '2020-02-05T12:06:11.758Z'}   
2795  {'$date': '2020-01-18T14:38:26.777Z'}   
3343  {'$date': '2020-01-17T14:57:55.899Z'}   
6973  {'$date': '2020-02-14T10:40:08.946Z'}   

                                            companyDesc  \
75    abbvie is a global, research-based biopharmace...   
626   astrazeneca is global innovation-driven biopha...   
1137  roche molecular systems, inc., a molecular dia...   
1184  following its acquisition of aetna in late 201...   
1394  covance inc. provides early-and late-stage ser...   
1478  eurofins scientific is an international life s...   
2135  icon is a global provider o

So here we've got bottom 10 pharmaceutical companies in terms of growth
</br>
**Note - The growth metric defined by me is based on the idea that the newer company with large company size in terms of manpower has good growth while an old company with small company manpower hasn't developed much. So the founded date of the company is in multiplication in numerator with the manpower of the comapny and 100000 in the denominator is used to scale the values to a much suitable range.**
<br>
<font color='blue'>So here we have completed the solution to our problem statement 2</font>

## Bonus Exercise
Present any insights from the data that you found interesting. (this exercise carries bonus points, but isn’t necessary)

#### Some observations
First of all when we look at the companyName column we observe that it is in the form of a list. So what can we infer from that.
The observation that i drew from it was that there are abbreviated names of a company present in the dataset like [informix, informix software india .]. Also it might happen that a company name is present with it's different county office designations like [tyco fire security india, tyco international]. Also there are cases like [carpe diem, carpediem] where the company name is same but with different pronunciation breakdowns.

In [55]:
unique_TYPE = df['TYPE'].dropna(axis=0)
unique_TYPE = unique_TYPE.value_counts()
unique_TYPE

Product-based    4465
Service-based    1632
Name: TYPE, dtype: int64

From above we can observe that most of the companies in out dataset are Product-based, let's see what percentage

In [58]:
(unique_TYPE[0]/sum(unique_TYPE))*100

73.23273741184188

So an astonishingly high 73% comapnies in our dataset are Product based

In [60]:
serie5 = df['domain_groups']
list5 = list(serie5)

In [65]:
count = 0
not_nan_count = 0
for i in list5:
    if str(type(i)) != "<class 'dict'>":
        pass
    elif len(i) > 1:
        count+=1
        not_nan_count += 1
    else:
        not_nan_count += 1
    

In [66]:
(count/not_nan_count)*100

4.630049931911031

So here we can make another revelation that despite many comapnies having multiple tags attatched to them but when it comes to the presence of a comany in a domain_group, only 4.63% companies have shown actual presence in in more than one domain group.

**Note - The above revelation is dependent upon my personal understanding of the data column domain_groups. Due to lack of information about the features of the dataset, this is how i have intepreted the column data.**

### Declaration:
The above work is my original without any tinge of any sort of malpractise involved. This is a work of hours of my dedication and hardwork. The above work is submitted to SPOTTABL. I do not have any rights over the data that has been used in this project.

THANK YOU :)